### PR TITLE
[persistence] Support for limit and offset filter on get all endpoints

### PR DESCRIPTION
**Pull request by shounakmk219**
#### Issue(s)

https://cortexdata.atlassian.net/browse/TE-1428

#### Description

Ability to get a subset of data using the `limit` and `offset` filter on get all entities endpoints.

Example : `http://localhost:8080/api/anomalies?isChild=false&limit=5&offset=10`
This will return only the next five anomaly entries by skipping the initial 10 entries.
Currently the default ordering is based on the ids (actual order in the table).

This will work across all entity endpoints.

### CODE REVIEW

File: thirdeye-integration-tests/src/test/java/ai/startree/thirdeye/PaginationTest.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+/**
+ * Use this for testing the bits related to pagination
+ */
+public class PaginationTest {
```
[the remark this answers is not in the document]
Comment (suvodeep-pyne) [reply thirdeye-integration-tests/src/test/java/ai/startree/thirdeye/PaginationTest.java:43]: Great to have an end to end test. We also need a test in the persistence layer itself. that would be super helpful

**Review by suvodeep-pyne — COMMENTED** [review]
LGTM. Please address comments

File: thirdeye-persistence/src/main/java/ai/startree/thirdeye/datalayer/dao/GenericPojoDao.java [hunk 2]
```diff
@@ -304,9 +304,11 @@ private <E extends AbstractDTO> List<E> filter(final Class<E> pojoClass,
 
   @SuppressWarnings("unchecked")
   public <E extends AbstractDTO> List<E> filter(final DaoFilter daoFilter) {
-    requireNonNull(daoFilter.getPredicate(),
-        "If the predicate is null, you can just do "
-            + "getAll() which doesn't need to fetch IDs first");
+    if (daoFilter.getLimit() == null) {
```
[the remark this answers is not in the document]
Comment (shounakmk219) [reply thirdeye-persistence/src/main/java/ai/startree/thirdeye/datalayer/dao/GenericPojoDao.java]: Done

File: thirdeye-integration-tests/src/test/java/ai/startree/thirdeye/PaginationTest.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    return apis.stream()
+        .map(ThirdEyeCrudApi::getId)
+        .collect(Collectors.toList());
+  }
+
+  @Test
+  public void testResponseSizeWithLimitFilter() {
+    final int limit = 20;
+    final int totalAnomalies = 100;
+    final List<AnomalyApi> anomalies = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (shounakmk219) [reply thirdeye-integration-tests/src/test/java/ai/startree/thirdeye/PaginationTest.java]: done

File: thirdeye-integration-tests/src/test/java/ai/startree/thirdeye/PaginationTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (shounakmk219) [reply thirdeye-integration-tests/src/test/java/ai/startree/thirdeye/PaginationTest.java]: Added separate persistence tests

File: thirdeye-integration-tests/src/test/java/ai/startree/thirdeye/PaginationTest.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    returnedAnomalies = response.readEntity(ANOMALY_LIST_TYPE);
+    assertThat(returnedAnomalies.size()).isEqualTo(totalAnomalies);
+  }
+
+  @Test(dependsOnMethods = "testResponseSizeWithLimitFilter")
+  public void testResponseWithLimitAndOffsetFilters() {
+    final int limit = 40;
+
+    List<AnomalyApi> page1 = getAllWithLimitAndOffset(limit, 0);
+    List<Long> allPagesIds = apisToIds(page1);
```
[the remark this answers is not in the document]
Comment (shounakmk219) [reply thirdeye-integration-tests/src/test/java/ai/startree/thirdeye/PaginationTest.java]: Sure

File: thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/dao/TestGenericPojoDao.java [hunk 5]
```diff
@@ -16,31 +16,48 @@
...
     dao = MySqlTestDatabase.sharedInjector().getInstance(GenericPojoDao.class);
+    for(int i=0; i<TOTAL_ANOMALIES; i++) {
+      dao.create(anomaly());
+    }
   }
 
   @AfterClass(alwaysRun = true)
   public void afterClass() {
     dao.deleteByPredicate(Predicate.NEQ(NAME, "null"), DataSourceDTO.class);
+    dao.deleteByPredicate(Predicate.NEQ(BASE_ID, 0), AnomalyDTO.class);
```
Comment (suvodeep-pyne) [inline thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/dao/TestGenericPojoDao.java:60]: what is this about?

Comment (suvodeep-pyne) [inline thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/dao/TestGenericPojoDao.java:60]: Please add a comment here

**Review by suvodeep-pyne — DISMISSED** [review]
LGTM

File: thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/dao/TestGenericPojoDao.java (hunk 5 above)
Comment (shounakmk219) [reply thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/dao/TestGenericPojoDao.java:60]: Is just cleans up the entries created during the tests

Comment (shounakmk219) [reply thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/dao/TestGenericPojoDao.java:62]: Added

**Pull request merged**